internal/providers/gitlab/registration.go

```diff
@@ -0,0 +1,172 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitlab
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/stacklok/minder/internal/entities/properties"
+	"github.com/stacklok/minder/internal/util/ptr"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// RegisterEntity implements the Provider interface
+func (c *gitlabClient) RegisterEntity(
+	ctx context.Context, entType minderv1.Entity, props *properties.Properties,
+) (*properties.Properties, error) {
+	if !c.SupportsEntity(entType) {
+		return nil, errors.New("unsupported entity type")
+	}
+
+	upstreamID := props.GetProperty(properties.PropertyUpstreamID).GetString()
+	if upstreamID == "" {
+		return nil, errors.New("missing upstream ID")
+	}
+
+	whprops, err := c.createWebhook(ctx, upstreamID)
+	if err != nil {
+		// There is already enough context in the error message
+		return nil, err
+	}
+
+	return props.Merge(whprops), nil
+}
+
+// DeregisterEntity implements the Provider interface
+func (c *gitlabClient) DeregisterEntity(
+	ctx context.Context, entType minderv1.Entity, props *properties.Properties,
+) error {
+	if !c.SupportsEntity(entType) {
+		return errors.New("unsupported entity type")
+	}
+
+	upstreamID := props.GetProperty(properties.PropertyUpstreamID).GetString()
+	if upstreamID == "" {
+		return errors.New("missing upstream ID")
+	}
+
+	hookID := props.GetProperty(RepoPropertyHookID).GetString()
+	if hookID == "" {
+		return errors.New("missing hook ID")
+	}
+
+	if err := c.deleteWebhook(ctx, upstreamID, hookID); err != nil {
+		// There is already enough context in the error message
+		return err
+	}
+
+	return nil
+}
+
+func (c *gitlabClient) createWebhook(ctx context.Context, upstreamID string) (*properties.Properties, error) {
+	createHookPath, err := url.JoinPath("projects", upstreamID, "hooks")
```

Comment:
we might want to split the JoinPath into a common function in case we change the path here but would forget to change the path in deleteWebhook